[KohGuanZeh] iP

KohGuanZeh opened the pull request:
# Duke Zeh

> I'm not writing it down to remember it _later_, I'm writing it down to remember it **now**. - [Field Notes](https://fieldnotesbrand.com/)

Duke Zeh is a task management application based on [Project Duke](https://nus-cs2103-ay1920s1.github.io/website/se-book-adapted/projectDuke/index.html)🔥.

## Command List
- `todo <task-name>`: Adds a new `ToDo` item.
- `deadline <task-name> \by <dd-MM-yyyy HH:mm>`: Adds a new `Deadline` task.
- `event <task-name> \from <dd-MM-yyyy HH:mm> \to <dd-MM-yyyy HH:mm>`: Adds a new `Event` item.
-  `mark <task-number>`: Marks task at index `<task-number>` as done.
- `unmark <task-number>`: Marks task at index `<task-number>` as not done.
- `delete <task-number>`: Deletes task at index `<task-number>`.
- `find <task-name>`: Finds task with matching keywords of `<task-name>`.
- `bye`: Exits the program.

## Levels Implemented
- [x] Level-0: Renaming Bot
- [x] Level-1: Greet, Echo, Exit
- [x] Level-2: Add, List
- [x] Level-3: Mark as Done
- [x] Level-4: ToDos, Events, Deadlines
- [x] Level-5: Handle Errors
- [x] Level-6: Delete
- [x] Level-7: Save
- [x] Level-8: Dates and Times
- [x] Level-9: Find
- [x] Level-10: GUI

## Running Duke Zeh
1. Download the repository [here](https://github.com/KohGuanZeh/ip).
2. Run `./gradlew clean build` from root folder.
3. Locate `duke.jar` from `builds/lib`.
4. Run `java -jar duke.jar`.

## Old Code Snippet of Duke Zeh
The following code was when Duke Zeh (formally known as Duke Ted) ran with CLI.
```java
public class Duke {
    // File directory of stored data.
    private static final String FILE_DIRECTORY = "data";
    // File name of stored data.
    private static final String FILE_NAME = "duke.txt";

    public static void main(String[] args) {
        Ui ui = new Ui();
        TaskList taskList = new TaskList();
        Storage storage = null;
        try {
            storage = new Storage(Duke.FILE_DIRECTORY, Duke.FILE_NAME);
            taskList.loadTasks(storage.load(), ui);
        } catch (IOException e) {
            ui.showError("Error in creating or accessing data file. Exiting...");
            return;
        }

        ui.showGreeting();

        boolean isQuit = false;
        while (!isQuit) {
            try {
                String input = ui.readline();
                ui.showLine();
                Command command = Parser.parseInput(input);
                command.run(taskList, ui, storage);
                isQuit = command.isExit();
            } catch (CommandException e) {
                ui.showError(e.getMessage());
            } catch (IOException e) {
                ui.showError("An error has occurred with the save file. Exiting...");
                isQuit = true;
            } finally {
                ui.showLine();
            }
        }
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

KohGuanZeh committed:
Update greeting and exit

KohGuanZeh committed:
Add echo feature, refactor greet and exit

KohGuanZeh committed:
Add store and list messages feature and refactor

Chatbot commands are now exclusively lower case.

This is to keep behaviour consistent in supporting future commands.

Looping behaviour is no longer separated to a function.

This is to allow for easier branching of behaviour as more commands are
supported. Future iterations can look into creating a separate chatbot
class and instance separate from the main() function.

KohGuanZeh committed:
Add Task class

KohGuanZeh committed:
Add TaskList class

KohGuanZeh committed:
Refactor Duke and TaskList

Duke:
- Remove all previous functions and replace with an echo() function.
- Standard messages are now stored as constants.

TaskList:
- markTask() and unmarkTask() now return String as output feedback.

Duke is refactored as many functions serve to only print a message.
It is redundant to create many functions to print to output.
echo() serves as a way to print output through the chatbot.

TaskList functions should return a message that will be sent as
feedback for the users. This is to make the chatbot feel more
responsive and users are aware of actions executed by chatbot.

KohGuanZeh committed:
Add different task types

ToDo: Task with no datetime attached.

Deadline: Task with a specified due date.

Event: Task with a specified start datetime and end datetime.

- User input to add tasks are now changed.
- TaskList addTask() function is slightly modified to immediately take
  in a created task.

KohGuanZeh committed:
Add automated text UI testing

KohGuanZeh committed:
Add error handling and refactor task creation

TaskException: Exceptions for when configuring tasks.

UnknownCommandException: Exception when user input unknown command.

Tasks are now created through factory methods from their respective
tasks. Tasks are also now created based on input directly after given
command.

This is to allow TaskExceptions to be thrown exclusively by Tasks
instead of the main program.

KohGuanZeh committed:
Change path of files

Prefer to run files from project root folder.

KohGuanZeh committed:
Add delete task

Also add such that adding and deleting tasks now outputs total number of
tasks left in the list.

Errors for delete task are also accounted for.

KohGuanZeh committed:
Update to enum commands

KohGuanZeh committed:
Add save load feature

KohGuanZeh committed:
Merge branch 'branch-Level-7'

KohGuanZeh committed:
Change String datetime to LocalDateTime

KohGuanZeh committed:
Merge branch 'branch-Level-8'

KohGuanZeh committed:
Add Ui class

Duke program is handling all tasks.

It is good to have greater separation of concerns to reduce coupling.

Ui class helps separate printing of outputs.

KohGuanZeh committed:
Add Storage class

Class is not yet used in program.

KohGuanZeh committed:
Refactor code base

Code is tightly coupled with no clear separations of handling.

Command class are now introduced together with Parser and Storage
to better separate concerns. Additionally, TaskException is now renamed
as CommandException to better represent the exception.

KohGuanZeh committed:
Merge branch 'branch-A-MoreOOP'

KohGuanZeh committed:
Fix bug for unmark, mark and delete

Input was not trimmed before parsing, causing commands to fail.

KohGuanZeh committed:
Merge branch 'branch-A-MoreOOP'

KohGuanZeh committed:
Organize code to respective package

KohGuanZeh committed:
Merge branch 'branch-A-Packages'

KohGuanZeh committed:
Merge remote-tracking branch 'origin/add-gradle-support'

KohGuanZeh committed:
Add gradle and unit test

KohGuanZeh committed:
Edit style of code

KohGuanZeh committed:
Add missing JavaDocs

KohGuanZeh committed:
Add find feature

KohGuanZeh committed:
Merge branch 'branch-A-JavaDoc'

KohGuanZeh committed:
Merge branch-A-CodingStandard

KohGuanZeh committed:
Merge branch 'branch-Level-9'

KohGuanZeh committed:
Fix missing import

KohGuanZeh committed:
Update README

KohGuanZeh committed:
Add JavaFX and refactor code

Important general changes:
1. Ui class is removed in favour of GUI with JavaFX.
2. Entry point of program has changed to Launcher.java instead of Duke.java

Command:
- Commands now return the String for response after calling command.

TaskList:
- loadTasks() now returns number of tasks that failed to load from
  improper formatting.

Minor changes:
- Renamed chatbot from Ted to Duke Zeh.

KohGuanZeh committed:
Edit chatbot response

Add more line breaks where necessary such that formatting on GUI is
cleaner.

KohGuanZeh committed:
Merge branch 'branch-Level-10'

KohGuanZeh committed:
Update README.md

jiahui0309 commented on src/main/java/duke/command/ToDoCommand.java:
```diff
@@ -0,0 +1,36 @@
+ * @author KohGuanZeh
+ */
+public class ToDoCommand extends Command {
+    private final String toDoDescription;
```
Final variables can be written in screaming snake case. Perhaps consider `TODO_DESCRIPTION`?

jiahui0309 commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,50 @@
+public class EventCommand extends Command {
+    private final String eventDescription;
+    private final LocalDateTime eventFrom;
+    private final LocalDateTime eventTo;
```
Author can consider using screaming snake case for final variables.

jiahui0309 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,119 @@
+    public static Command parseInput(String input) throws CommandException {
+        input = input.stripLeading();
+
+        if (input.equals("bye")) {
+            return new ByeCommand();
+        } else if (input.equals("list")) {
+            return new ListCommand();
+        } else if (input.startsWith("todo")) {
+            return new ToDoCommand(input.substring(4));
+        } else if (input.startsWith("deadline")) {
+            return new DeadlineCommand(input.substring(8));
+        } else if (input.startsWith("event")) {
+            return new EventCommand(input.substring(5));
+        } else if (input.startsWith("mark")) {
+            return new MarkCommand(input.substring(4));
+        } else if (input.startsWith("unmark")) {
+            return new UnmarkCommand(input.substring(6));
+        } else if (input.startsWith("delete")) {
+            return new DeleteCommand(input.substring(6));
+        } else if (input.startsWith("find ")) {
+            // Account for space since parameters will not be trimmed.
+            return new FindCommand(input.substring(5));
+        }
```
I like how you used PascalCase for class names, and that I can easily infer the purpose of the class based on the class names.

jiahui0309 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,119 @@
+            Task task = null;
+            String description = "";
+            boolean marked = Integer.parseInt(tokens[1].trim()) > 0;
+            switch (tokens[0].trim()) {
+            case "T":
```
I like how you used `break;` for every `case` statement and that there is no indentation for `case` clauses.

jiahui0309 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+        return "There are now " + this.taskList.size() + " tasks in the list.";
+    }
+
+    /**
+     * Sets the task at the specified index as done and returns a task completion message.
+     *
+     * @param index Index of task as displayed.
+     * @return Task completion message.
+     */
```
Perhaps you can consider adding `@throws` in the Javadoc as there is an exception thrown in the header.

jiahui0309 reviewed (commented):
LGTM! The variable names are very clearly defined, making It easy to read. I noticed that there were several instances of final variables not being in screaming snake cases. I like how you incorporated Javadocs for most of your methods. :)

KohGuanZeh committed:
Update JavaDocs missing exceptions

KohGuanZeh committed:
Merge branch 'master' of https://github.com/KohGuanZeh/ip

KohGuanZeh committed:
Add asserts

Program does not use any assertions, and have no validation of states
before running further operations.

This can cause the program to run with errors and cause unintended
consequences.

Assertions are added at various points of the script, mainly
Parser.java, Storage.java and Main.java.

This is done for early program termination in the case that important
assertions are not met.

KohGuanZeh committed:
Refactor Parser

Local variable tokens is reused to continually split the String
to get LocalDateTime information from splitting.

Can cause wrong information to be parsed if tokens as tokens are
overwritten at different points.

New local variables are created to extract LocalDateTime information.

This will prevent confusion and avoid splitting on wrong data.

In the future should avoid cases of reusing variables.

KohGuanZeh committed:
Refactor ByeCommand message

ByeCommand returns a magic String literal. The String is then recognized
by MainWindow.java to know when to exit the program.

Typo or change in message can cause program to work incorrectly as it as
a functional component apart from printing to output.

Storing the message as a constant under ByeCommand will allow for better
consistency and improve readability.

In the future should avoid magic literals, especially if it affects
logic in the code.

KohGuanZeh committed:
Refactor TaskList functions using index

Same check on index is is done before retrieving or removing task from
markTask, unmarkTask and deleteTask.

Check condition uses some form of complex expression repeatedly.

Extract out the index bounds check into separate functions that get
and removes tasks from taskList. Also introduce new TaskListException
to handle such exceptions.

KohGuanZeh committed:
Merge pull request #1 from KohGuanZeh/branch-A-Assertions

Add asserts

KohGuanZeh committed:
Merge pull request #2 from KohGuanZeh/branch-A-CodeQuality

Merge branch-A-CodeQuality

KohGuanZeh committed:
Refactor Commands and Parser

Commands attempt to parse input to get parameters to create tasks within
constructor. Constructor is also reusing input parameter and reassigning
a trimmed input to overwrite input.

Same code is duplicated in parseData, with the exception of errors
thrown.

Parser class can include functions to parse input directly into
a task and still falls under responsibility of Parser.

This provides better encapsulation in turn providing better readability.

In the future, should better utilize Parser class to parse data from one
type to another.

KohGuanZeh committed:
Refactor Task toDataString function

Reduce complex logic in written code.

Changed StringBuilder to append String before returning as it looks
cleaner.

KohGuanZeh committed:
Merge pull request #3 from KohGuanZeh/branch-A-CodeQuality

Merge branch-A-CodeQuality

KohGuanZeh committed:
Add Priority feature

Task have no separation in priorities.

Add different priority levels for user to better manage their tasks.
There are 3 priority levels, low, high and none. Default priority value
is none and there will be no indication beside listed task.

KohGuanZeh committed:
Merge branch 'branch-C-Priority'

KohGuanZeh committed:
Create gradle.yml

KohGuanZeh committed:
Fix JavaDoc comment

KohGuanZeh committed:
Change filename for stored data

KohGuanZeh committed:
Rename images

KohGuanZeh committed:
Add credit for reuse code

KohGuanZeh committed:
Add UI screenshot

KohGuanZeh committed:
Update README.md

KohGuanZeh committed:
Update README
